test/core/surface/channel_init_test.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/lib/surface/channel_init.h"
+
+#include <algorithm>
+#include <map>
+#include <string>
+
+#include "absl/strings/string_view.h"
+#include "gtest/gtest.h"
+
+#include "src/core/lib/channel/channel_stack.h"
+#include "src/core/lib/channel/channel_stack_builder_impl.h"
+#include "test/core/util/test_config.h"
+
+namespace grpc_core {
+namespace {
+
+const grpc_channel_filter* FilterNamed(const char* name) {
+  static auto* filters =
+      new std::map<absl::string_view, const grpc_channel_filter*>;
+  auto it = filters->find(name);
+  if (it != filters->end()) return it->second;
+  return filters
+      ->emplace(name,
+                new grpc_channel_filter{nullptr, nullptr, nullptr, 0, nullptr,
+                                        nullptr, nullptr, 0, nullptr, nullptr,
+                                        nullptr, nullptr, name})
+      .first->second;
+}
+
+std::vector<std::string> GetFilterNames(const ChannelInit& init,
+                                        grpc_channel_stack_type type,
+                                        const ChannelArgs& args) {
+  ChannelStackBuilderImpl b("test", type, args);
+  init.CreateStack(&b);
+  std::vector<std::string> names;
+  for (auto f : b.stack()) {
+    names.push_back(f->name);
+  }
+  return names;
+}
+
+TEST(ChannelInitTest, Empty) {
+  EXPECT_EQ(GetFilterNames(ChannelInit::Builder().Build(), GRPC_CLIENT_CHANNEL,
+                           ChannelArgs()),
+            std::vector<std::string>());
+}
+
+TEST(ChannelInitTest, OneClientFilter) {
+  ChannelInit::Builder b;
+  b.RegisterFilter(GRPC_CLIENT_CHANNEL, FilterNamed("foo"));
+  auto init = b.Build();
```

Comment:
Shouldn't this be failing because there is no terminal filter?

Reply:
now that there's rules updated a bunch of tests here... worse to read tests, but probably safer